### PR TITLE
Support for PaloAlto FWAAS.

### PR DESCRIPTION
**Pull request by rayalaman**
@ashutosh-mishra @njagadish @chenchenWang @songole 

This is a pull request for palo alto firewall driver support 

added Feature:
orchestration:
- added support in nfp node driver to pick paloalto firewall service
- no hotplug support.
- no sharing support.

configuration:
- create_interface: create ip on the static interface
- create_route: creates pbr rule to route traffic from provider to stitching 
- create_firewall: creates firewall rules
- delete_firewall: deletes firewall rules

### CODE REVIEW

**Comment by rayalaman** [comment]
FYI - we are depending on a paloalto related python library called "python-pan" which can be installed using either pip or through git clone: https://github.com/kevinsteves/pan-python. Was not sure on how we handling driver dependency packages. @songole suggested when we are doing rpms we can install this library when the configurator rpm is installed. for now we can put it in run.sh in the configurator image.

**Comment by ashutosh-mishra** [comment]
@rayalaman - Discuss with @DhuldevValekar for how to install this dependency

**Comment by rayalaman** [comment]
@ashutosh-mishra yes we decided to do a git clone of the pan-os when when we do image builder of the configurator vm.

File: gbpservice/nfp/configurator/drivers/firewall/paloalto/paloAlto_fw_driver.py [hunk 1]
```diff
@@ -0,0 +1,926 @@
...
+        Issues REST call to service VM for configuration of interfaces.
+
+        :param context: neutron context
+        :param kwargs: a dictionary of firewall rules and objects
+        send by neutron plugin
+
+        Returns: SUCCESS/Failure message with reason.
+
+        """
+        time.sleep(60)
```
Comment (ashutosh-mishra) [inline gbpservice/nfp/configurator/drivers/firewall/paloalto/paloAlto_fw_driver.py]: Why blind sleep ?
If we need sleep, add proper comment why we need this, and put it in constants file.

File: gbpservice/nfp/configurator/drivers/firewall/paloalto/paloAlto_fw_driver.py [hunk 2]
```diff
@@ -0,0 +1,926 @@
...
+        vm_mgmt_ip = kwargs.get('mgmt_ip') 
+        self.check_auto_commit_status(hostname=vm_mgmt_ip)
+        
+        # apply staic ip to stitching interface
+        stitching_interface_index = kwargs.get('stitching_interface_index')
+        stitching_interface_name = "ethernet1/%d" % \
+                                        (stitching_interface_index-1)
+        stitching_ip = kwargs.get('stitching_ip')
+        stitching_ip = ("%s/24" % stitching_ip)
```
Comment (ashutosh-mishra) [inline gbpservice/nfp/configurator/drivers/firewall/paloalto/paloAlto_fw_driver.py]: which is the correct stmnt

File: gbpservice/nfp/configurator/drivers/firewall/paloalto/paloAlto_fw_driver.py (hunk 1 above)
Comment (rayalaman) [inline gbpservice/nfp/configurator/drivers/firewall/paloalto/paloAlto_fw_driver.py]: even after the service agent comes up, there is a auto commit job that take a certain amount to come up, for the time being we could not find a way to check for the completion of the job. hence the sleep is required here. we will add the sleep time to the constants file.

File: gbpservice/nfp/configurator/drivers/firewall/paloalto/paloAlto_fw_driver.py [hunk 3]
```diff
@@ -0,0 +1,926 @@
...
+                                      const.UPDATE_SECURITY_RULE, resp)
+                raise Exception(err)
+
+            if self.analyze_response("UPDATED the configuration to Service VM",
+                                     resp, const.UPDATE_SECURITY_RULE)\
+                                        == const.STATUS_ERROR:
+                return const.STATUS_ERROR
+        """
+
+        # commite security rule
```
Comment (ashutosh-mishra) [inline gbpservice/nfp/configurator/drivers/firewall/paloalto/paloAlto_fw_driver.py]: Is update a No OP, as we are not doing anything for update ?
If yes, then why do we need to do commit ?

File: gbpservice/nfp/configurator/drivers/firewall/paloalto/paloAlto_fw_driver.py [hunk 4]
```diff
@@ -0,0 +1,926 @@
...
+        Returns: UPDATED/ERROR status
+
+        """
+
+        vm_mgmt_ip = self._get_firewall_attribute(firewall)
+        msg = ("Initiating UPDATE request.")
+        LOG.info(msg)
+    
+        """
+        # update security rule
```
Comment (ashutosh-mishra) [inline gbpservice/nfp/configurator/drivers/firewall/paloalto/paloAlto_fw_driver.py]: Remove commented code

**Comment by ashutosh-mishra** [comment]
@njagadish - changes are fine

**Comment by njagadish** [comment]
Closing this as it got merged as part of 03 June PR

**Pull request closed without merging**